File: src/main/java/part4/Lim3085.java
package part4;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Lim3085 {
    private static int max = 0;
    private static int n;

    public static void main(String[] args) throws IOException {
        final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        final char[][] candies = new char[n][n];

        for (int i = 0; i < n; i++) {
            candies[i] = br.readLine().toCharArray();
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (j + 1 < n && candies[i][j] != candies[i][j + 1]) {
                    change(candies, i, j, i, j + 1);
                    search(candies);
                    change(candies, i, j, i, j + 1);
                }

                if (i + 1 < n && candies[i][j] != candies[i + 1][j]) {
                    change(candies, i, j, i + 1, j);
                    search(candies);
                    change(candies, i, j, i + 1, j);
                }
            }
        }

        System.out.println(max);
    }

    private static void change(final char[][] candies, final int i, final int j, final int i2, final int j2) {
        final char temp = candies[i][j];
        candies[i][j] = candies[i2][j2];
        candies[i2][j2] = temp;
    }

    private static void search(final char[][] candies) {
        for (int i = 0; i < n; i++) {
            searchHorizontal(candies, i, 0);
            searchVertical(candies, 0, i);
        }
    }

    private static void searchHorizontal(final char[][] candies, final int i, final int j) {
        int length = 1;

        for (int k = j + 1; k < n; k++) {
            if (candies[i][j] == candies[i][k]) {
                length++;
            } else {
                searchHorizontal(candies, i, k);
                break;
            }
        }

        if (length > max) {
            max = length;
        }
    }

    private static void searchVertical(final char[][] candies, final int i, final int j) {
        int length = 1;

        for (int k = i + 1; k < n; k++) {
            if (candies[i][j] == candies[k][j]) {
                length++;
            } else {
                searchVertical(candies, k, j);
                break;
            }
        }

        if (length > max) {
            max = length;
        }
    }
}

File: src/main/java/boj/part7/Lim3085.kt
package boj.part7

import java.io.BufferedReader
import java.io.InputStreamReader

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()
    val board = Array(n) { CharArray(n) }
    repeat(n) {
        board[it] = br.readLine().toCharArray()
    }

    var max = 0
    for (i in 0 until n) {
        for (j in 0 until n - 1) {
            swapHorizontal(board, i, j)
            val result = searchMax(board, n)
            if (result > max) {
                max = result
            }
            swapHorizontal(board, i, j)
        }
    }

    for (i in 0 until n - 1) {
        for (j in 0 until n) {
            swapVertical(board, i, j)
            val result = searchMax(board, n)
            if (result > max) {
                max = result
            }
            swapVertical(board, i, j)
        }
    }

    println(max)
}

private fun swapHorizontal(board: Array<CharArray>, i: Int, j: Int) {
    val temp = board[i][j]
    board[i][j] = board[i][j + 1]
    board[i][j + 1] = temp
}

private fun swapVertical(board: Array<CharArray>, i: Int, j: Int) {
    val temp = board[i][j]
    board[i][j] = board[i + 1][j]
    board[i + 1][j] = temp
}

private fun searchMax(board: Array<CharArray>, n: Int): Int {
    var max = 0

    for (i in 0 until n) {
        for (j in 0 until n) {
            val hResult = searchHorizontal(board, n, i, j)
            val vResult = searchVertical(board, n, i, j)
            max = maxOf(hResult, vResult, max)
        }
    }

    return max
}

private fun searchHorizontal(board: Array<CharArray>, n: Int, i: Int, j: Int): Int {
    var result = 1

    for (k in j + 1 until n) {
        if (board[i][j] != board[i][k]) {
            break
        }
        result++
    }

    return result
}

private fun searchVertical(board: Array<CharArray>, n: Int, i: Int, j: Int): Int {
    var result = 1

    for (k in i + 1 until n) {
        if (board[i][j] != board[k][j]) {
            break
        }
        result++
    }

    return result
}
